### DISCUSSION TITLE
Different header file/define for each test

### DISCUSSION THREAD

**Discussion by mrrb** [ThrowTheSwitch/Ceedling#597] [Q&A]
Hi!

I am trying to test an embedded library that I'm working on.
Some parts of the code are conditionally hidden with a ```#ifdef DEFINE_NAME ... #else ... #endif``` block, and I have a config header file (ex. lib_config.h) where the lib user can select (comment/uncomment #defines) which functionalities want to use.

How can I use a different config header file for each test, so I can try all the different lib combinations?

Thank you! 😃

**Comment by art-of-dom** [comment]
I have a hackyish way to do this for an internal library that I do a little bit differently just with defining specific variables, but I think it will work for you. You can leverage environmental defines and use them to change defines in your code.  For your specific case, I'd suggest using the env variables to include additional headers like so.

```
#ifdef CASE1
#include "lib_config_case1.h"
#elif defined(CASE2)
#include "lib_config_case2.h"
...
#endif
```

---------------------------------

Once you have the variables sorted out that way, add it to your defines and you can write a script to set the variable/overwrite it. Because it's kind of intensive to do both, I'll just share a paired down version of my project.yml to give you ideas. In mine, I'm defining two sets of variables like a 2d matrix, but you should be able to get away with one. 

```
:defines:
  :commmon: &common_defines
    [
      TEST, LOGGER_VERBOSITY=3, 
      CMOCK_MEM_DYNAMIC, CCM_USE_PREDEFINED_VALUES
    ]
  :test:
    - *common_defines
    - UNITY_SUPPORT_TEST_CASES
    - ${ROLE_DEFINE-CONTROLLER}
    - ${SECURITY_DEFS-ALL_SECURITY}
  :test_preprocess:
    - *common_defines
    - UNITY_SUPPORT_TEST_CASES
    - ${ROLE_DEFINE-CONTROLLER}
    - ${SECURITY_DEFS-ALL_SECURITY}
    - TEST_PP
```


```
#!/bin/bash
set -e

export ROLE_DEFINE="CONTROLLER"
export SECURITY_DEFS="ALL_SECURITY"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="SECURITY_0"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="SECURITY_2"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="NO_SECURITY"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export ROLE_DEFINE="AUX"
export SECURITY_DEFS="ALL_SECURITY"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="SECURITY_0"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="SECURITY_2"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

export SECURITY_DEFS="NO_SECURITY"
printenv ROLE_DEFINE
printenv SECURITY_DEFS
rake clean clobber test:all

```

**Reply by mrrb** [reply]
Thank you for the reply, I will try it out!
